# Project 3 - Web APIs & NLP by Chenhao Andy Xu

In [33]:
import pandas as pd
import numpy as np

import requests
from bs4 import BeautifulSoup
import time

## 1. Download Data

### 1.1. Get Data for Piano

In [38]:
base_url = 'https://api.pushshift.io/reddit/search/'

In [39]:
params = {
    'subreddit': 'piano', 
    'size': 100, 
}

res = requests.get(base_url + 'submission/', params = params)

In [40]:
res.status_code

521

In [41]:
data = res.json()

JSONDecodeError: Expecting value: line 1 column 1 (char 0)

In [ ]:
piano_all = pd.DataFrame(data['data'])
piano_all

In [ ]:
pd.DataFrame(data['data']).columns

In [ ]:
piano = piano_all[['subreddit', 'title', 'selftext', 'created_utc']]
piano

In [ ]:
for i in range(40):
    time.sleep(1)
    params = {'subreddit': 'piano', 'size': 100, 'before': piano[-1:]['created_utc']}
    res = requests.get(base_url + 'submission/', params = params)
    # res.status_code
    data = res.json()
    df_all = pd.DataFrame(data['data'])
    df = df_all[['subreddit', 'title', 'selftext','created_utc']]
    piano = pd.concat([piano, df])

In [ ]:
piano

In [ ]:
piano.to_csv('./data/piano.csv')

In [ ]:
piano.isna().sum()

In [ ]:
piano.shape

### 1.2. Get Data for Drum

In [ ]:
params = {
    'subreddit': 'drum', 
    'size': 100, 
}

res = requests.get(base_url + 'submission/', params = params)

In [ ]:
res.status_code

In [ ]:
data = res.json()

In [ ]:
drum_all = pd.DataFrame(data['data'])
drum_all

In [ ]:
drum = drum_all[['subreddit', 'title', 'selftext', 'created_utc']]
drum

In [ ]:
for j in range(39):
    time.sleep(2)
    params = {'subreddit': 'drum', 'size': 100, 'before': drum[-1:]['created_utc']}
    res = requests.get(base_url + 'submission/', params = params)
    # res.status_code
    data = res.json()
    df_all = pd.DataFrame(data['data'])
    df = df_all[['subreddit', 'title', 'selftext','created_utc']]
    drum = pd.concat([drum, df])

In [ ]:
drum

In [ ]:
drum.to_csv('./data/drum.csv')

In [ ]:
drum.isna().sum()

In [ ]:
drum.shape

In [ ]:
drum.info()

I had unexpected issue when I got data from subreddit. The piano part was fine. But the same code for drum always showed some error (although I was still getting decent amount of data from it). I talked to Jeff and he said the code was fine. I don't know where the issue is. 

One more update: now, the res.json() even showed 521 not 200. I don't know where I made mistake. 